pythonPackages.qutip: init at 4.5.2

mcncm opened the pull request:
* Does not use OpenMP by default
* May be overly conservative in disabling python versions older than 3.5
* Note that build time is a bit long, in part due to tests. It might be
  preferable to disable these in the future.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

This is a fairly popular quantum simulation package that it would be nice to have easy access to.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- ~~[ ] Tested execution of all binary files (usually in `./result/bin/`)~~ (N/A)
- [x] Built the package with `nix-shell -p 'python$X.withPackages (ps: [ ps.qutip ])'`, for `X`={`36`,`37`,`38`}, and imported the module in a repl of the respective Python version.
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mcncm committed:
pythonPackages.qutip: init at 4.5.2

* Does not use OpenMP by default
* May be overly conservative in disabling python versions older than 3.5
* Note that build time is a bit long, in part due to tests. It might be
  preferable to disable these in the future.

SuperSandro2000 commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,65 @@
+{ stdenv
+, lib
```
Only use stdenv and stdenv.lib in all other places.

mcncm replied:
Hi Sandro,

First-time contributor here. I take that to mean that `lib` is unused? Thanks for pointing that out! I admit to roughly pattern-matching against other pythonPackages, and easily could have done a thing or two without knowing why.

While you're taking a look at this PR, I have two other questions if that's all right:
1. The bot `grahamcofborg-eval` failed with the message, "This PR does not cleanly list package outputs after merging." What does that mean? What can I do to fix it?
2. It appears that this bot also doesn't like the maintainer name (`mcncm`, _i.e._ me) isn't in scope. Should it be an attribute of `maintainers`? Why is this building successfully for me if there's an undefined variable?

Thank you!

SuperSandro2000 replied:
> I take that to mean that lib is unused?

It is not really unused but the one place it is used it can easily be replaced with ``stdenv.lib``. Tbh it is just a nitpick to have less inputs. It does not really change anything functional wise.

> It appears that this bot also doesn't like the maintainer name (mcncm, i.e. me) isn't in scope. Should it be an attribute of maintainers? Why is this building successfully for me if there's an undefined variable?

You need to add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix. The contributing guide requires to do that in a separate commit. The commit message should look like ``maintainers: add name``.

> The bot grahamcofborg-eval failed with the message, "This PR does not cleanly list package outputs after merging." What does that mean? What can I do to fix it?

If you follow the details link you can read the log which complains about the missing maintainer.

```
undefined variable 'mcncm' at 
```

https://gist.github.com/GrahamcOfBorg/02f8ab994b78fc43abd018687bec154c

mcncm replied:
Hi Sandro,

Thanks for the elaboration. I've removed the dependence on `lib`, added myself as a maintainer, and modified the derivation slightly to add openmp support to qutip.

mcncm committed:
maintainers: add mcncm

mcncm committed:
pythonPackages.qutip: build with openmp

* Overridden `buildPhase` to build with openmp support
* As this does not appear to work with the Pypi version, which lacks
  certain .cpp source files, `fetchPypi` has been replaced with
  `fetchGitHub`.
* The correct tests have been added, but disabled, as they add a
  considerable amount of time after an already quite long build phase.

SuperSandro2000 commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,69 @@
+    matplotlib
+    cython
+  ];
+
+  nativeBuildInputs = [ numpy scipy cython pytest setuptoolsBuildHook ];
```
Those dependencies are probably only needed in one *buildInputs.

SuperSandro2000 replied:
```suggestion
```

SuperSandro2000 commented:
Result of `nixpkgs-review pr 103412` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python37Packages.qutip</li>
    <li>python38Packages.qutip</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,69 @@
+{ stdenv
+, python
+, buildPythonPackage
+, pythonOlder
+, fetchFromGitHub
+, setuptoolsBuildHook
+, gcc
+, hypothesis
+, numpy
+, scipy
+, matplotlib
+, cython
+, pytest
+}:
```
```suggestion
{ stdenv
, python
, buildPythonPackage
, pythonOlder
, fetchFromGitHub
, setuptoolsBuildHook
, hypothesis
, numpy
, scipy
, matplotlib
, cython
, openmp
, sysctl
}:
```

SuperSandro2000 commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,69 @@
+      --replace "numpy>=1.12" "numpy" \
+      --replace "scipy>=1.0" "scipy"
+  '';
+
```
```suggestion

  nativeBuildInputs = [ setuptoolsBuildHook ];

  buildInputs = stdenv.lib.optionals stdenv.isDarwin [ openmp ];

```

SuperSandro2000 commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,69 @@
+  buildPhase = ''
+  ${python.interpreter} setup.py --with-openmp bdist_wheel
+  '';
+
+  doCheck = false;
```
```suggestion

  doCheck = false;
```

Tests fail one darwin and linux.

jonringer replied:
please leave a comment saying why tests werent ran

SuperSandro2000 commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+  doCheck = false;
+
+  checkInputs = [ hypothesis ];
```
```suggestion
  checkInputs = [ hypothesis sysctl ];
```

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6116,6 +6116,8 @@ in {
 
   queuelib = callPackage ../development/python-modules/queuelib { };
 
+  qutip = callPackage ../development/python-modules/qutip { };
```
```suggestion
  qutip = callPackage ../development/python-modules/qutip {
    inherit (pkgs.llvmPackages) openmp;
  };
```

SuperSandro2000 commented:
Result of `nixpkgs-review pr 103412` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python37Packages.qutip</li>
    <li>python38Packages.qutip</li>
  </ul>
</details>

jonringer commented on pkgs/development/python-modules/qutip/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+  enableParallelBuilding = true;
+
+  buildPhase = ''
+  ${python.interpreter} setup.py --with-openmp bdist_wheel
+  '';
```
```suggestion
  setupPyGlobalFlags = "--with-openmp";
```

jonringer reviewed (commented):
Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the fix-up commits.

 `git rebase -i` is a powerful command which achieves this, I created a small video demonstrating it's use [here](https://www.youtube.com/watch?v=OR8Q-4PLalc). A more indepth text tutorial can be found [here](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

jonringer commented:
to allow tests:
```
diff --git a/pkgs/development/python-modules/qutip/default.nix b/pkgs/development/python-modules/qutip/default.nix
index 14a13683272..b8a104d8cc7 100644
--- a/pkgs/development/python-modules/qutip/default.nix
+++ b/pkgs/development/python-modules/qutip/default.nix
@@ -43,17 +43,14 @@ buildPythonPackage rec {

   enableParallelBuilding = true;

-  buildPhase = ''
-  ${python.interpreter} setup.py --with-openmp bdist_wheel
-  '';
-
-  doCheck = false;
+  setupPyBuildFlags = "--with-openmp";

   checkInputs = [ hypothesis ];

   checkPhase = ''
     runHook preCheck
     pushd dist
+    HOME=$TMPDIR
     ${python.interpreter} -c 'import qutip.testing as qt; qt.run()'
     popd
     runHook postCheck
```

SuperSandro2000 commented:
@mcncm ping

FabianGD commented:
Closing this, as a newer version of the package is provided by #144102
